Add PlatformNativeTokenProvider using the AuthenticationServices framework

dcow opened the pull request:
Add a platform native token provider that uses the authentication services framework to do oauth2. Here it is in action:

https://user-images.githubusercontent.com/547926/142566963-93a2e752-b10e-4197-b87a-4b12bb55b52e.mov

The first run is the browser provider. The second run is the native flow. 

Obviously this pr needs work before merging (prints, copy/paste cleanup, fixup commits, etc.) but I'm opening it now to see if there's any interest.

dcow committed:
oauth2: add PlatformNativeTokenProvider

The platform provider uses the native AuthenticationServices framework
to handle the authenitication request. This means that the terminal
redirect will get sent back into the app instead of landing the user
on a page in the browser. This provider is macOS/iOS only.

timburks commented on Sources/OAuth2/PlatformNativeTokenProvider.swift:
```diff
@@ -0,0 +1,180 @@
+// Copyright 2019 Google LLC. All Rights Reserved.
```
Nit: please make this 2021

timburks reviewed (commented):
Thanks for starting this! I'm not current on the AuthenticationServices framework, but I can take a deeper look or find someone else to help review this whenever you say it's ready for final review and merge.

dcow commented:
@timburks to brief you: the quick pitch for Apple's AuthServices framework is that for native client use cases, the framework allows you to share auth credentials with the system browser solely for the purpose of authenticating a user in an oauth-style flow. Normally, an app's WKWebViews don't share cookies with the system browser. Apple added the SFSafariViewController as a solution but people quickly abused it to generally steal credentials and track people, so the shared cookie jar was removed after iOS 10. To fill the void for legitimate use cases, Apple introduced the ASWebAuthenticationSession which displays a remote safari view in your application that shares auth cookies with the system browser but is limited to machinery and a UX tailored for doing auth flows (the only data you're allowed to exfiltrate comes via the terminal redirect). It's a neat solution. The framework also has types and functionality to implement "enterprise" SSO with a "third party" provider as well as "first party" support for login with apple, credential autofill, etc. It might be interesting to look at some of the SSO stuff and see if it could replace any of the hand rolled types here, but I digress. 

Anyway, I'm ready to start working through a review. My main questions are:

* Does this belong in this library? It uses an Apple framework so it's not pure Swift. It makes sense to me to include it with the general distribution for people targeting mac platforms because of the UX advantage, but there's a question of how to package it up and present it so that this library continues to work as expected on non mac platforms (it might already work just fine due to the `@available` checks, but I haven't tested).
* I need to test whether the refresh token logic pulled over from the BrowserTokenProvider actually works on iOS. Reading it I didn't see a problem but it would be good to verify.
* Would it be preferable to structure the api as a throwing function similar to the browser provider and use dispatch semaphores to block the caller? The problem I see with doing that is that the the ASWebAuthenticationSession probably needs to be started from the main thread since it interacts with UI. However that's just a guess I haven't tried it and don't have the source for the framework to audit myself. I structured the api in this PR in a way that I believe will feel natural to iOS uses with a callback that receives an optional Token if the auth flow succeeds. I think we might be able to add a little more error reporting to the callback as remove the prints (assuming those should be removed).
* I'll probably add a shared "exchange" implementation that takes a Code and returns a Token so it can be used in both the browser provider and this new one.

dcow commented:
@timburks alright I fixed up the error handling and removed prints. This should be ready for a proper review now.

dcow commented on Sources/OAuth2/PlatformNativeTokenProvider.swift:
```diff
@@ -0,0 +1,156 @@
+import AuthenticationServices
+
+struct NativeCredentials: Codable, CodeExchangeInfo {
+  let clientID: String
```
Not sure if there are cases where a client secret is used for a native flow like this. Probably wouldn't hurt to support it in the credentials structure but it's not used when creating credentials for a native app in the cloud console so I left it out.

dcow replied:
Rather, I've used providers that *do* like to use a client secret even for native app flows where it's not secret, but this code is google specific and I don't *think* there are scenarios where Google expects/uses one. Someone that knows Google better than I should confirm, though.

timburks replied:
If this works for your application, it sounds good to me. Pinging @leisurehound for any other comments.

dcow commented on Sources/OAuth2/PlatformNativeTokenProvider.swift:
```diff
@@ -14,7 +14,9 @@
 import Dispatch
 import Foundation
-import AuthenticationServices
+#if canImport(AuthenticationServices)
```
This plus the `@available` attribute on the class together don't seem to be enough to prevent compilation on Linux. I'll have to look into other options (maybe macro-ing out the entire file/class). I'm not super versed in slicing up code to build properly across both Apple and Linux platforms so clever suggestions are welcome.

timburks replied:
Thanks for looking out for this (I think we want to keep our Ubuntu builds working). I think it would be fine to wrap the entire file contents in an #if directive, as described [here](https://developer.apple.com/forums/thread/653437) - Since it builds on MacOS, you probably want more than `#if os(iOS)`, maybe `#if os(iOS) || os(macOS)` (or whatever is the correct syntax for this).

leisurehound replied:
+1 to ensuring this flow works on macOS as well as iOS.  I can see use cases on macOS clients that this would make convenient.

timburks commented:
@dcow Thanks again for contributing this! I haven't been actively using Swift for the last few years, so I won't try to address style, but perhaps @leisurehound or @algal will want to step in and advise.

To your questions:

1. I do think this belongs in this library (meaning it's a welcome addition). I'd like to keep our Linux builds working, but I think we can guard the new code to make that possible.
2. Regarding throwing function / semaphores: this library predates most of the async work in Swift, and I would expect it to *not* match current idioms on iOS. I trust your judgement on the best structure for iOS users.

So as far as I'm concerned, I think this is good to merge as soon as the Ubuntu CI test passes.

leisurehound commented:
Can you add an example of how its used to be explicit for any users?

dcow commented:
Where would you like the example?

timburks commented:
There's a Sources/Examples directory but those are built with the Swift Package Manager. How about a new Examples directory at the top level with one subdirectory containing an iOS example?

timburks reviewed (approved):
LGTM - @dcow before merging, I'll wait to hear from you about an example to address @leisurehound's request. We can include that in this PR or merge this and take that in a subsequent PR, whichever you prefer.

dcow commented:
> LGTM - @dcow before merging, I'll wait to hear from you about an example to address @leisurehound's request. We can include that in this PR or merge this and take that in a subsequent PR, whichever you prefer.

I have the example I had been using cleaned up and worked into this repo. Everything works fine except the native flow on macOS just isn't opening the browser after I select `Continue`:
<img width="434" alt="Screen Shot 2021-11-22 at 1 56 55 PM" src="https://user-images.githubusercontent.com/547926/142934153-896710aa-d56c-4eee-8e95-cbe3d13cbc5d.png">

No logs, and the callback isn't being called with an error or anything. I'm running macOS 12.0.1 and don't really have the time at the moment to do a ton of investigation if it doesn't become obvious why it isn't working. The example is still valuable to see how things are supposed to work using the Browser and Native providers so it might make sense for me to push it and see if anyone else experiences the issue and if we can't solve it just merge it and check back later.

dcow committed:
example: add macOS/iOS native demo

Add examples of the embedded browser and the AuthenticationServices
flows in an xcodeproject. The macOS flow isn't working at the moment.

leisurehound commented:
Its likely an entitlements issue on the Mac side.

On Nov 22, 2021 at 1:01:42 PM, David ***@***.***> wrote:

> LGTM - @dcow <https://github.com/dcow> before merging, I'll wait to hear
> from you about an example to address @leisurehound
> <https://github.com/leisurehound>'s request. We can include that in this
> PR or merge this and take that in a subsequent PR, whichever you prefer.
>
> I have the example I had been using cleaned up and worked into this repo.
> Everything works fine except the native flow on macOS just isn't opening
> the browser after I select Continue:
> [image: Screen Shot 2021-11-22 at 1 56 55 PM]
> <https://user-images.githubusercontent.com/547926/142934153-896710aa-d56c-4eee-8e95-cbe3d13cbc5d.png>
>
> No logs, and the callback isn't being called with an error or anything.
> I'm running macOS 12.0.1 and don't really have the time at the moment to do
> a ton of investigation if it doesn't become obvious why it isn't working.
> The example is still valuable to see how things are supposed to work using
> the Browser and Native providers so it might make sense for me to push it
> and see if anyone else experiences the issue and if we can't solve it just
> merge it and check back later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/googleapis/google-auth-library-swift/pull/55#issuecomment-975914363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABNLO2O6OYI525ZGH3LLYDUNKVTNANCNFSM5ILITEFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

dcow commented:
@leisurehound I just added the example project. I was messing with entitlements and whatnot for a bit to no avail. I will try again some more later tonight.

dcow commented:
I'll squash this down into two commits, 1 for the provider and one for the example project.

leisurehound commented:
Am trying to load/run the example, but being pulled into this project only recently and new to OAuth as well I was able to get it to run but with some observations. 

I do think the credentials help could be a bit more explicit for users new to this area.  Specifically, I tried to replace the JSON with what was created on the console and they keys did not match up, rather than just pull out the client ID and secret from the downloaded JSON which I finally figured out.

It might also be helpful to alert users to enable the gmail api, otherwise the response has PERMISSION DENIED error perhaps allowing them to conclude it didn't work.

Lastly, on first run on macOS I did get the native prompt which took me to the Google auth page, but doing so again is indeed silent.  Is there a short circuit here because it has a recent token that is still valid? Is it possible it will 'work' again after 1hr?

dcow commented:
> Am trying to load/run the example, but being pulled into this project only recently and new to OAuth as well I was able to get it to run but with some observations.
> 
> I do think the credentials help could be a bit more explicit for users new to this area. Specifically, I tried to replace the JSON with what was created on the console and they keys did not match up, rather than just pull out the client ID and secret from the downloaded JSON which I finally figured out.
> 
> It might also be helpful to alert users to enable the gmail api, otherwise the response has PERMISSION DENIED error perhaps allowing them to conclude it didn't work.

I agree it's rather confusing for someone new to the whole thing to set it up. However I kinda scoped the commentary to an audience that is familiar with gauth/oauth2 so as not to end up with this being an entire tutorial. Regardless, I'll add more instructions and clarify. Never hurts to over-communicate.

> Lastly, on first run on macOS I did get the native prompt which took me to the Google auth page, but doing so again is indeed silent. Is there a short circuit here because it has a recent token that is still valid? Is it possible it will 'work' again after 1hr?

What doesn't make sense to me about this is what happens if I lose the token and refresh token? I'd need to get another one. On iOS (if you run that target) you'll see that every time you press the button you get a new token.

dcow committed:
fixup: add more gauth setup details

* And move logic to its own file.

dcow commented:
@leisurehound I updated the explanation on how to get the example working. Let me know if it's sufficient.

dcow committed:
example: add token caching

dcow commented:
> What doesn't make sense to me about this is what happens if I lose the token and refresh token? I'd need to get another one. On iOS (if you run that target) you'll see that every time you press the button you get a new token.

I think I figured out what was happening. It seems all the links were getting sent to Safari Technology Preview. For some reason it wasn't opening and I had no idea it should be. I opened it up today for something else and now a bunch of the OAuth flows are launching sequentially. Seems to be a macOS issue unrelated to the implementation here.

leisurehound commented:
> > What doesn't make sense to me about this is what happens if I lose the token and refresh token? I'd need to get another one. On iOS (if you run that target) you'll see that every time you press the button you get a new token.
> 
> I think I figured out what was happening. It seems all the links were getting sent to Safari Technology Preview. For some reason it wasn't opening and I had no idea it should be. I opened it up today for something else and now a bunch of the OAuth flows are launching sequentially. Seems to be a macOS issue unrelated to the implementation here.

Sorry for the delay here, I will get back to this today or tomorrow.  Thanks for our persistence on it.

dcow commented:
One thing my colleague noticed is that there's no token refresh logic only token save/load (caching) logic. Is that something we should perhaps add in a followup?

leisurehound commented:
I'm getting the following exception trying to run the example and press the native login:

```
requesting new token
2021-12-02 09:37:10.302659-0800 GAuth-macOS[75119:21819788] [AuthenticationSession] The provided scheme is not valid. A scheme should not include special characters such as ":" or "/".
2021-12-02 09:37:10.306882-0800 GAuth-macOS[75119:21819788] [General] The provided scheme is not valid. A scheme should not include special characters such as ":" or "/".
```

Is there something else to configure beyond enabling the mail api?

dcow commented:
> Is there something else to configure beyond enabling the mail api?

There shouldn't be. Can you paste the value you're using for `callback_scheme` in the example project credentials json?

leisurehound commented:
Sorry, pilot error.  Pulling your latest changes I did not apply all the changes from my working client post pull correctly.

dcow commented:
@leisurehound any further comments from you I can address before we wind down for the week? Would love to get this merged.

leisurehound commented:
Sorry, I was away from the office.  The code looks good.  I've built the example successfully and see the browser version working, but am also seeing the native exhibiting the behavior you first described.  I inferred with the subsequent changes you resolved that, is that not the case?

dcow commented:
@leisurehound the browser flow works all the time on macOS/iOS. The native flow works without fuss on iOS. On macOS, if you run the app from Xcode and encounter a native flow that dies after pressing okay when the system prompts for credential use, try restarting Safari. When Safari relaunches you will see the auth window pop up (you may see a bunch of them from previous runs the first time--they all get queued up) and, after granting access, the flow redirects to the app and completes successfully as expected. My conclusion is that everything is working fine and the behavior we're seeing is simply some quirk with how the system/browser handles these requests for apps launched from Xcode (e.g. possibly some system application registry that's getting updated lazily and only read on-launch by Safari).

leisurehound commented:
Thanks for the PR @dcow and sorry for the delay.

dcow commented:
No worries! Thanks for the review. I may have another PR shortly with some token refresh logic (=
